var/spack/repos/builtin/packages/dyninst/package.py

```diff
@@ -53,6 +53,7 @@ class Dyninst(CMakePackage):
 
     patch('stat_dysect.patch', when='+stat_dysect')
     patch('stackanalysis_h.patch', when='@9.2.0')
+    patch('v9.3.2-auto.patch', when='@9.3.2')
```

Comment:
I would also restrict the patch by GCC version, whatever versions didn't support the newer syntax only.

Reply:
I wasn't sure how to write that, that is, how to combine the @9.3.2
with the %gcc syntax inside when().

Suppose the line is between 4.x and 5.x.  How would I write this?

But I don't think the restriction is all that necessary.  The patch is
six occurrences of replace:

```
for(const auto& t: typesByName) {
   ...
}
```

with:

```
for(auto t = typesByName.begin(); t != typesByName.end(); ++t) {
   ...
}
```

So, it really is a no-op.

Reply:
The only reason I would restrict it is that this doesn't seem like the kind of patch you would submit upstream. And since it hasn't been vetted by the developers, I don't want to always apply it.

> Suppose the line is between 4.x and 5.x. How would I write this?

If you want to patch 4.x and 5.x (inclusive), it would look like:
```python
    patch('v9.3.2-auto.patch', when='@9.3.2 %gcc@4:5')
```
The syntax is identical to the spec syntax you use on the command-line when specifying how to build a package.